### ISSUE TITLE
NullReferenceException in GetPlaceholderInformationAsyncHandler

### ISSUE THREAD

**Issue by kewillford** [Microsoft/VFSForGit#893]
Version 1.0.19060.1 logged a NullReferenceException

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at GVFS.Common.Tracing.JsonTracer.MarkAndLogListenerFailure(EventListener failedListener, String errorMessage)
   at GVFS.Common.Tracing.JsonTracer.WriteEvent(String eventName, EventLevel level, Keywords keywords, EventMetadata metadata, EventOpcode opcode)
   at GVFS.Common.Tracing.JsonTracer.RelatedEvent(EventLevel level, String eventName, EventMetadata metadata, Keywords keyword)
   at GVFS.Common.Tracing.JsonTracer.RelatedError(EventMetadata metadata, String message, Keywords keywords)
   at GVFS.Platform.Windows.WindowsFileSystemVirtualizer.GetPlaceholderInformationAsyncHandler(CancellationToken cancellationToken, BlobSizesConnection blobSizesConnection, Int32 commandId, String virtualPath, UInt32 desiredAccess, UInt32 shareMode, UInt32 createDisposition, UInt32 createOptions, UInt32 triggeringProcessId, String triggeringProcessImageFileName) in E:\A\_work\586\s\GVFS\GVFS.Platform.Windows\WindowsFileSystemVirtualizer.cs:line 803
```

**Comment by wilbaker** [comment]
Not sure if this is related, but `failedListeners` is a regular `Dictionary` and not a `ConcurrentDictionary` and as best I can tell it can get modified on multiple threads without any locking.

**Comment by mjcheetham** [comment]
Given the stack it appears we got into the actual `Dictionary.Insert` method, meaning the `NullReferenceException` occurred inside of the dictionary. Checking the .NET Framework codebase (not the .NET Core one since that's considerably different.. there's no `private void Insert(key,value,add)` method) it appears the `key` was _not_ `null` or else the exception we'd get would be `ArgumentNullException`.

I think you're correct @wilbaker that this is a concurrency issue. The only other possible place a reference could be `null` would be the internal `buckets` field which is initialised lazily for the first time inside a `null` check.

The current usage pattern of the in `JsonTracer` of the `failedListeners` field is a 'contains check + add/remove' which should be coverable by the `TryAdd` and `TryRemove` methods on `ConcurrentDictionary`.

However I do also notice that there's no locking or though to concurrency with the _`List<EventListener> listeners`_ field adding listeners, or writing messages to them in a loop.
There is no `ConcurrentList`, only a bag. We could just do a `lock` around access to the `listeners` field too. Although given we've not hit a problem (have we?) around that field before, do we need to also worry about this?

**Comment by wilbaker** [comment]
>I do also notice that there's no locking or thought to concurrency with the List<EventListener> listeners field adding listeners

That's a great point.  

I suspect when it comes to adding `listeners` we've been OK because we always add the listeners right away or before starting up any other threads that use the tracer.  Having another set of eyes confirm that would be good 😄   

I agree we could improve the safety here with something like a `ConcurrentBag`.

> or writing messages to them in a loop

On this point I think we need to check the implementation of each listener.  

_LogFileEventListener_

This one is okay because its `writer` is synchronized:

```
this.writer = StreamWriter.Synchronized(new StreamWriter(this.logFile));
```

_PrettyConsoleEventListener_

This one is okay because it has a lock:

```
// It's necessary to do a lock here because this can be called in a multi-threaded
// environment and we want to make sure that ForegroundColor is restored correctly.
lock (consoleLock)
{
```

_TelemetryDaemonEventListener_

**I think we have a problem here**, as it looks like multiple threads could try to write to `this.pipeClient` at the same time:

```
var buffer = Encoding.UTF8.GetBytes(message);
this.pipeClient.Write(buffer, 0, buffer.Length);
```

And if two threads hit an exception, one could get a null exception when it tries to dispose `this.pipeClient`

```
catch (Exception)
{
    // We can't send this message for some reason (e.g., broken pipe); we attempt no recovery or retry
    // mechanism and drop this message. We will try to recreate/connect the pipe on the next message.
    this.pipeClient.Dispose();
    this.pipeClient = null;
```

**Comment by wilbaker** [comment]
@jrbriggs @mjcheetham I will create a follow up issue for the thread other thread safety issues raised by @mjcheetham and prepare a fix for 147 for the original `MarkAndLogListenerFailure` issue mentioned in this work item.

**Comment by wilbaker** [comment]
Fixed in #901 and #907
